README.md
pymapper
========

Cross-platform gui for libmapper

[![pymapper](http://acaia.ca/~tiago/images/pymapper.png "pymapper")](https://github.com/tiagovaz/pymapper)


debian/changelog
pymappergui (0.1-1) unstable; urgency=low

  * Initial release. (Closes: #000000)

 -- Tiago Bortoletto Vaz <tiago@debian.org>  Tue, 11 Nov 2014 19:06:15 -0500


testing/test.py
from pyo import *
from pyolib._wxwidgets import ControlSlider, BACKGROUND_COLOUR
import wx

import mapper

class MyFrame(wx.Frame):
    def __init__(self, parent, title, size, colour="#DDDDE7"):
        wx.Frame.__init__(self, parent, -1, title=title, size=size)
        sizer = wx.BoxSizer(wx.VERTICAL)
        self.dev = mapper.device("test")

        pos_y = 20 # y position for labels & slides
        s_label = 0 # integer which will generate incremented labels
        signal_id = 0 # this will be used to set the proper mapper signal

        #TODO: generate list of outputs and get their 'real' names to create the labels - duhhh
        for i in range(10): # generating lables & slides list
            self.slider = MyControlSlider(self, -1000, 1000, 0, pos=(45, pos_y), backColour=colour, signal_id=signal_id)
            sizer.Add(wx.StaticText(self, -1, str(s_label)+".", pos =(20, pos_y)), 0, wx.ALIGN_CENTRE|wx.ALL, 5)
            s_label += 1
            pos_y += 25
            signal_id += 1

        # mapper poll
        self.timer = wx.Timer(self)
        self.Bind(wx.EVT_TIMER, self.OnTimer)
        self.timer.Start(milliseconds=10, oneShot=False)

    def OnTimer(self, evt):
        self.dev.poll(1)

class MyControlSlider(ControlSlider):
    def __init__(self, parent, minvalue, maxvalue, init=None, pos=(0,0), size=(200,16), log=False,
                 integer=False, powoftwo=False, backColour=None, signal_id=None):

        ControlSlider.__init__(self, parent=parent, minvalue=minvalue, maxvalue=maxvalue, init=init,
                               pos=pos, size=size, log=log, outFunction=self.setMapper, integer=integer,
                               powoftwo=powoftwo, backColour=backColour)

        self.signal_id = signal_id
        self.mon = mapper.monitor()
        self.dev = self.GetParent().dev
        self.output = self.dev.add_output("/"+str(self.signal_id), 1, 'f', None, -1000, 1000)

        while not self.dev.ready():
            self.dev.poll(10)
            self.mon.poll()

    def setMapper(self, value):
        self.output.update(value)

app = wx.App(False)
frame = MyFrame(None, title="testing pymapper", size=(280, 300))
frame.Show()
app.MainLoop()



testing/Audio.py
from pyo import *
from random import *

class DarkAtmosphere:
    def __init__(self, p1=.2054, p2=[.3966, .3933], chaos=.025, feedback=.5, mul=.5):
        self.r1 = Rossler(pitch=p1, chaos=.4038, stereo=True, mul=mul, add=.2)
        self.r2 = Rossler(pitch=p2, chaos=.0769, mul=self.r1*.8)
        self.amp = Fader(fadein=5, fadeout=5, dur=0)
        self.d = Delay(self.r2, feedback=.5, mul=self.amp)

    def setFeedback(self, feedback):
        self.d.feedback = feedback
        return self

    def setPitch(self, p1, p2):
        self.r1.pitch = p1
        self.r2.pitch = p2
        return self

    def setChaos(self, chaos):
        self.r1.chaos = chaos
        return self

    def out(self):
        self.d.out()
        self.amp.play()
        return self

    def tail(self):
        self.amp.stop()

    def sig(self):
        return self.d

    def stop(self):
        self.d.stop()
        self.r1.stop()
        self.r2.stop()
        return self

    def getOut(self):
        return self.amp

    def setInput(self, x, fadetime=.001):
        self.input.setInput(x, fadetime)

class MyDelay:
    """A sensor controls delay/reverb of an audio stream
       - audio1 being an Input() signal
       - audio2 being a sensor output (audio stream) from 0 to 1
       - audio3 being a sensor output (audio stream) from 40 to 800
    """
    #TODO: no mess, take fadein/out as argument
    def __init__(self, audio1, audio2=.01, audio3=40, mul=.7):
        self.lfd = Sine([.4,.3], mul=.2, add=.5)
        self.supersaw = SuperSaw(freq=[ audio3*uniform(0.94, 1.04) for i in range(10) ], detune=self.lfd, bal=0.7, mul=.1)
        self.r = Freeverb(self.supersaw, size=.94, damp=.87, bal=.9, mul=1)
        self.d2 = Delay(self.r, delay=audio2, feedback=uniform(0.95, 1), mul=.4)
        self.d2.ctrl()
        self.d3 = Delay(self.d2, delay=[.05, .1, .25, .5], feedback=.25, mul=.4)
        self.d3.ctrl()
        #self.c = Chorus(audio1, depth=[1.5,1.6], feedback=0.5, bal=0.5)
        self.r2 = Freeverb(audio1, size=.74, damp=.87, bal=.5, mul=1)
        self.d4 = Delay(self.r2, delay=audio2, feedback=uniform(0.95, 1), mul=.6)
        self.d4.ctrl()
        #self.mix = Freeverb(Mix(self.d2 + self.d3 + self.r2 + self.d4, mul=.7), size=.34, damp=.37, bal=.9, mul=.9)
        self.amp = Fader(fadein=20, fadeout=20)
        self.mix = Mix(self.d2 + self.d3 + self.r2 + self.d4, mul=self.amp*mul).out()

    def setSawFreq(self, freq):
        self.supersaw = SuperSaw(freq=[ freq*uniform(0.94, 1.04) for i in range(10) ], detune=self.lfd, bal=0.7, mul=.1)

    def setDelay(self, delay):
        self.d4.delay = delay

    def out(self):
        self.amp.play()
        return self

    def stop(self):
        self.amp.stop()
        return self

    def getOut(self):
        return self.amp

    def setInput(self, x, fadetime=.001):
        self.input.setInput(x, fadetime)

class SmoothNoise():
    def __init__(self, dur=1.3):
        self.amp = Fader(fadein=.1, fadeout=.01, dur=dur)
        self.noise = PinkNoise(self.amp*.01).mix(2).out()

    def setDur(self, dur):
        self.amp.dur = dur
        return self

    def out(self):
        self.amp.play()
        return self

    def stop(self):
        self.amp.stop()
        return self

    def getOut(self):
        return self.amp

    def setInput(self, x, fadetime=.001):
        self.input.setInput(x, fadetime)

class HighFreq():
    def __init__(self, freq=[11200, 11202], dur=.3):
        self.amp = Fader(fadein=.01, fadeout=.01, dur=dur)
        self.sine = SineLoop(freq=freq, mul=self.amp*.05).out()
        self.rev = Freeverb(self.sine, size=.84, damp=.87, bal=.9, mul=self.amp*.2).out()

    def setDur(self, dur):
        self.amp.dur = dur
        return self

    def out(self):
        self.amp.play()
        return self

    def stop(self):
        self.amp.stop()
        return self

    def getOut(self):
        return self.amp

#TODO: give random Randi freqs
class Wind():
    """Simple wind generator with Pyo"""
    def __init__(self, mul):
        # Brown noise to have less high frequency to cutoff
        self.noise = BrownNoise(mul=Randi(0.5,1))

        # This gives a factor for both the wind intensity and frequency. Intensity
        # and frequency go together in the two first filters in order to generate a
        # more natural wind. Q variation gives a kind of wind blows.
        self.j = Randi(1,1.4,Randi(.5,1))

        # Filter 1 - sort of continuos bass wind
        self.freqs1 = Randi(150, 300)
        self.q1 = Randi(1, 6)
        self.f1 = ButBP(self.noise, freq=self.freqs1, q=self.q1, mul=.2*self.j)

        # Filter 2 - main wind frequency with slow variation, following wind intensity
        self.freqs2 = [Randi(300, 400)*self.j for i in range(3)]
        self.q2 = Randi(40, 300)
        self.f2 = ButBP(self.noise, freq=self.freqs2, q=self.q2, mul=self.j*1.5)

        # Filter 3 - very high component, almost fixed in frequency
        self.freqs3 = [Randi(2990, 3000)*self.j for i in range(2)]
        self.q3 = Randi(1, 33)
        self.f3 = ButBP(self.noise, freq=self.freqs3, q=self.q3, mul=.01)

        # Filter 4 - the highest frequency wind component, also quite fixed
        self.freqs4 = [Randi(10000, 10100)*self.j for i in range(2)]
        self.q4 = Randi(10, 33)
        self.f4 = ButBP(self.noise, freq=self.freqs4, q=self.q4, mul=.01)

        # Player
        self.amp = Fader(fadein=5, fadeout=5)
        self.mix = Mix([self.f1, self.f2, self.f3, self.f4],voices=2, mul=self.amp*mul).out()

    def setAmp(self, amp):
        self.mix.mul = amp

    def out(self):
        self.amp.play()
        return self

    def stop(self):
        self.amp.stop()
        return self

    def getOut(self):
        return self.amp

class MyFM():
    def __init__(self, index=1, ratio=1, amp=.1, pitch=300):
        self.amp = amp
        self.pitch = pitch
        self.ratio = ratio
        self.index = index
        self.fm1 = FM(carrier=self.pitch, ratio=self.ratio, index=self.index, mul=self.amp*0.1).mix(1)
        self.fm2 = FM(carrier=self.pitch*1.01, ratio=self.ratio, index=self.index, mul=self.amp*0.1).mix(1)
        self.fm3 = FM(carrier=self.pitch*0.99, ratio=self.ratio, index=self.index, mul=self.amp*0.1).mix(1)
        self.fm4 = FM(carrier=self.pitch*1.04, ratio=self.ratio, index=self.index, mul=self.amp*0.1).mix(1)
        self.outmix = Mix([self.fm1, self.fm2, self.fm3, self.fm4], voices=2).out()

    def setRatio(self, r):
        self.ratio = r
        self.ratio = r
        self.ratio = r
        self.ratio = r

    def setIndex(self, index):
        self.index = self.index = self.index = self.index = index

    def setAmp(self, mul):
        self.fm1.mul = self.fm2.mul = self.fm3.mul = self.fm4.mul = mul

    def setAmp(self, pitch):
        self.pitch = pitch * random.uniform(0.97, 1.03)
        self.pitch = pitch * random.uniform(0.97, 1.03)
        self.pitch = pitch * random.uniform(0.97, 1.03)
        self.pitch = pitch * random.uniform(0.97, 1.03)

    def out(self):
        self.outmix.out()

    def stop(self):
        self.outmix.stop()


testing/audio.py
from pyo import *

class FreqMod:
    def __init__(self, carrier=150):
        self.ratio = SigTo(0.5, time=0.05, init=0.5)
        self.index = SigTo(5, time=0.05, init=5)
        self.fm1 = FM(carrier=carrier, ratio=self.ratio, index=self.index, mul=0.3)
        self.fm2 = FM(carrier=carrier*1.01, ratio=self.ratio, index=self.index, mul=0.3)
        self.fmst = Mix([self.fm1, self.fm2], voices=2)

    def out(self):
        self.fmst.out()
        return self

    def sig(self):
        return self.fmst

    def setRatio(self, x):
        xx = x * 1.999 + 0.001
        self.ratio.value = xx

    def setIndex(self, x):
        xx = x * 20
        self.index.value = xx

if __name__ == "__main__":
    s = Server(audio="jack").boot()
    fm = FreqMod(150).out()
    harm = Harmonizer(fm.sig(), transpo=2).out()
    s.gui(locals())


Resources/panels.py
import wx
import wx.gizmos
import mymapper
import math

class ConnectionsPanel(wx.Panel):
    def __init__(self, parent):
        wx.Panel.__init__(self, parent, style=wx.SUNKEN_BORDER)
        self.SetBackgroundColour("white")
        self.Bind(wx.EVT_PAINT, self.OnPaint)
        self.Bind(wx.EVT_SIZE, self.OnSize)
        self.my_mapper = mymapper.MyMapper()

    def OnSize(self, evt):
        wx.CallAfter(self.Refresh)

    def OnPaint(self, evt):
        self.DrawConnectionsLines()

    def Clear(self):
        self.dc.Clear()

    def DrawConnectionsLines(self):
        # loop source and dest trees to get y from each item connected
        conn_ids = []
        for c in self.my_mapper.getConnections():
            src = c["src_name"]
            dest = c["dest_name"]
            src_item_id = self.GetParent().GetParent().sources_panel.GetItemByLabel(src, self.GetParent().GetParent().sources_panel.GetRootItem())
            dest_item_id = self.GetParent().GetParent().destinations_panel.GetItemByLabel(dest, self.GetParent().GetParent().destinations_panel.GetRootItem())
            conn_ids.append((src_item_id, dest_item_id))
        w,h = self.GetSize()
        self.dc = wx.PaintDC(self)
        gc = wx.GraphicsContext_Create(self.dc)
        self.dc.Clear()
        gc.SetPen(wx.Pen("blue", 1))
        for conn in conn_ids:
            y1 = self.GetParent().GetParent().sources_panel.getItemPos(conn[0])
            y2 = self.GetParent().GetParent().destinations_panel.getItemPos(conn[1])
            # +30 to compensate for the header row
            l = []
            y1 = y1[1]+y1[3]/2+26
            y2 = y2[1]+y2[3]/2+26
            for i in range(w):
                mu = float(i)/w
                mu2 = (1-math.cos(mu*math.pi))/2
                l.append([i, y1*(1-mu2)+y2*mu2])
            gc.DrawLines(l)
        wx.CallAfter(self.Refresh)

class MyTreeList(wx.Panel):
    def __init__(self, parent, id, which):
        wx.Panel.__init__(self, parent, -1, style=wx.SUNKEN_BORDER)
        self.my_mapper = mymapper.MyMapper()

        # sources or destinations tree
        self.which = which

        # Create the tree
        self.tree = wx.gizmos.TreeListCtrl(self, style =
                                           wx.TR_DEFAULT_STYLE
                                           | wx.TR_FULL_ROW_HIGHLIGHT
                                           | wx.TR_HAS_BUTTONS + wx.TR_HIDE_ROOT,
                                           size=(530, 100))
        self.tree.Bind(wx.EVT_TREE_ITEM_ACTIVATED, self.OnActivate, self.tree)
        self.tree.Bind(wx.EVT_TREE_ITEM_EXPANDED, self.changed, self.tree)
        self.tree.Bind(wx.EVT_TREE_ITEM_COLLAPSED, self.changed, self.tree)
        self.tree.Bind(wx.EVT_TREE_SEL_CHANGED, self.OnItemChanged, self.tree)
        self.Bind(wx.EVT_SIZE, self.OnSize)

        self.devices_list = []

        ### expand all
        expand_icon = wx.ArtProvider.GetBitmap(wx.ART_UNDO, wx.ART_TOOLBAR, (16,16))
        expand_button = wx.BitmapButton(self, -1, expand_icon, (20, 20), (expand_icon.GetWidth()+10, expand_icon.GetHeight()+10))
        expand_button.SetToolTipString("This is a bitmap button.")
        self.Bind(wx.EVT_BUTTON, self.OnExpand, expand_button)

        box = wx.BoxSizer(wx.VERTICAL)
        box.Add(self.tree, 1)

        # create some columns
        self.tree.AddColumn("Devices & signals")
        self.tree.AddColumn("type")
        self.tree.AddColumn("lengh")
        self.tree.AddColumn("units")
        self.tree.AddColumn("minimum")
        self.tree.AddColumn("maximum")
        self.tree.SetMainColumn(0) # the one with the tree in it...
        self.tree.SetColumnWidth(0, 200)
        self.tree.SetColumnWidth(1, 50)
        self.tree.SetColumnWidth(2, 50)
        self.tree.SetColumnWidth(3, 50)
        self.tree.SetColumnWidth(4, 80)

        # load the tree nodes
        self.RefreshAll()

        # Expand the first level by default
        # FIXME: breaks if there's no item in the list
        #self.tree.ExpandAll(self.root)


    def OnItemChanged(self, evt):
        # show the connection expression and src/dest min/max in the toolbar
        src = self.GetParent().GetParent().sources_panel.GetSignalAddress()
        dest = self.GetParent().GetParent().destinations_panel.GetSignalAddress()
        #FIXME: gizmos module seems to have a bug here
        connection_data = self.my_mapper.getConnectionBySignalFullNames(src, dest)

        if connection_data is None:
            # disable controls
            self.GetParent().GetParent().mode_choice.Disable()
            self.GetParent().GetParent().mute_tool.Disable()
            self.GetParent().GetParent().expression_input.Disable()
            self.GetParent().GetParent().expression_y.Disable()
            self.GetParent().GetParent().arrow_range.Disable()
            self.GetParent().GetParent().arrow_range2.Disable()
            self.GetParent().GetParent().src_range_label.Disable()
            self.GetParent().GetParent().src_range_min.Disable()
            self.GetParent().GetParent().src_range_max.Disable()
            self.GetParent().GetParent().dest_range_label.Disable()
            self.GetParent().GetParent().dest_range_min.Disable()
            self.GetParent().GetParent().dest_range_max.Disable()
            self.GetParent().GetParent().expression_input.SetValue("")

            # set default values in the control panel
 #           self.GetParent().GetParent().expression_input.Clear()
            #for i in range(self.GetParent().GetParent().expression_input.GetCount()):
             #   self.GetParent().GetParent().expression_input.Delete(i)

            self.GetParent().GetParent().src_range_min.SetToDefaultValue()
            self.GetParent().GetParent().src_range_max.SetToDefaultValue()
            self.GetParent().GetParent().dest_range_min.SetToDefaultValue()
            self.GetParent().GetParent().dest_range_max.SetToDefaultValue()
        else:
            # enable controls
            self.GetParent().GetParent().mode_choice.Enable()
            self.GetParent().GetParent().mute_tool.Enable()
            if connection_data["mode"] == self.my_mapper.mo_expression:
                self.GetParent().GetParent().expression_input.Enable()
                self.GetParent().GetParent().expression_y.Enable()
            else:
                self.GetParent().GetParent().expression_input.Disable()
                self.GetParent().GetParent().expression_y.Disable()

            self.GetParent().GetParent().arrow_range.Enable()
            self.GetParent().GetParent().arrow_range2.Enable()
            self.GetParent().GetParent().src_range_label.Enable()
            self.GetParent().GetParent().src_range_min.Enable()
            self.GetParent().GetParent().src_range_max.Enable()
            self.GetParent().GetParent().dest_range_label.Enable()
            self.GetParent().GetParent().dest_range_min.Enable()
            self.GetParent().GetParent().dest_range_max.Enable()

            # set connection values in the control panel
            src_min = connection_data["src_min"]
            src_max = connection_data["src_max"]
            dest_min = connection_data["dest_min"]
            dest_max = connection_data["dest_max"]
            if connection_data["muted"] == 1:
                self.GetParent().GetParent().mute_tool.SetValue(True)
            elif connection_data["muted"] == 0:
                self.GetParent().GetParent().mute_tool.SetValue(False)

            # we need to match connection mode index with the choice index in the control panel:
            mode = int(connection_data["mode"])
            mode_choice_index = self.GetParent().GetParent().modes_list.index(self.my_mapper.modes_dict_rev[mode])
            self.GetParent().GetParent().mode_choice.SetSelection(mode_choice_index)

            self.GetParent().GetParent().expression_input.SetValue(connection_data["expression"].split('=')[1])
            self.GetParent().GetParent().src_range_min.SetValue(src_min)
            self.GetParent().GetParent().src_range_max.SetValue(src_max)
            self.GetParent().GetParent().dest_range_min.SetValue(dest_min)
            self.GetParent().GetParent().dest_range_max.SetValue(dest_max)

    def changed(self, event):
        self.GetParent().GetParent().redraw()

    def getAllItems(self):
        pass
#        print self.tree.GetItemText(self.root)

    def getItemPos(self, item):
        while not self.tree.IsVisible(item): # cool!
            item = self.tree.GetItemParent(item)
        return self.tree.GetBoundingRect(item)

    def getCurrentItem(self):
        return self.tree.GetCurrentItem()

    def OnActivate(self, event):
        self.NewConnection()
        self.GetParent().GetParent().connections_panel.DrawConnectionsLines()

    def NewConnection(self):
        # set link and connection with libmapper
        conns = self.my_mapper.getConnectionBySignalFullNames(self.GetParent().GetParent().sources_panel.GetSignalAddress(), self.GetParent().GetParent().destinations_panel.GetSignalAddress())
        if conns is None:
            self.my_mapper.setLink("/"+self.GetParent().GetParent().sources_panel.GetSignalAddress().split("/")[1], "/"+self.GetParent().GetParent().destinations_panel.GetSignalAddress().split("/")[1], {})
            self.my_mapper.Connect(self.GetParent().GetParent().sources_panel.GetSignalAddress(), self.GetParent().GetParent().destinations_panel.GetSignalAddress(), options={})
        else:
            self.my_mapper.Disconnect(self.GetParent().GetParent().sources_panel.GetSignalAddress(), self.GetParent().GetParent().destinations_panel.GetSignalAddress())

    def OnSize(self, event):
        self.tree.SetSize(self.GetSize())


    def AddTreeNodes(self, parentItem, items):
        for item in items:
            bg_color = 0
            if self.which == "destinations" and self.my_mapper.getInputsFromDevice(item):
                newItem = self.tree.AppendItem(parentItem, item)
                self.tree.SetItemBackgroundColour(newItem, "#D3D3D3")
                self.tree.SetItemBold(newItem)
                for i in self.my_mapper.getInputsFromDevice(item):
                    s_input = self.tree.AppendItem(newItem, i["name"])
                    if bg_color % 2 == 0:
                        self.tree.SetItemBackgroundColour(s_input, "#F5F5F5")
                    bg_color += 1
                    self.tree.SetItemText(s_input, i["type"], 1)
                    self.tree.SetItemText(s_input, str(i["length"]), 2)
                    self.tree.SetItemText(s_input, "", 3)
                    self.tree.SetItemText(s_input, str(i["min"]), 4)
                    self.tree.SetItemText(s_input, str(i["max"]), 5)

            elif self.which == "sources" and self.my_mapper.getOutputsFromDevice(item):
                newItem = self.tree.AppendItem(parentItem, item)
                self.tree.SetItemBackgroundColour(newItem, "#D3D3D3")
                self.tree.SetItemBold(newItem)
                for o in self.my_mapper.getOutputsFromDevice(item):
                    s_output = self.tree.AppendItem(newItem, o["name"])
                    if bg_color % 2 == 0:
                        self.tree.SetItemBackgroundColour(s_output, "#F5F5F5")
                    bg_color += 1
                    self.tree.SetItemText(s_output, o["type"], 1)
                    self.tree.SetItemText(s_output, str(o["length"]), 2)
                    self.tree.SetItemText(s_output, "", 3)
                    self.tree.SetItemText(s_output, str(o["min"]), 4)
                    self.tree.SetItemText(s_output, str(o["max"]), 5)

    def RefreshAll(self):
        # Add nodes (devices and their signals) to the trees
        # keep last devices list, compare with current and update if changed
        # same for removing item (DO NOT REBUILD ALL TREE)
        self.new_devices_list = []
        self.new_devices = self.my_mapper.getAllDevices()
        if self.new_devices != self.devices_list:
            self.tree.DeleteAllItems()
            # Hidden tree root node
            self.root = self.tree.AddRoot("")
            self.AddTreeNodes(self.root, self.new_devices)
            self.tree.Refresh()
            self.ExpandAll()
            self.devices_list = self.new_devices

    def OnExpand(self, event):
        self.ExpandAll()

    def ExpandAll(self):
        self.tree.ExpandAll(self.root)

    def GetRootItem(self):
        return self.tree.GetRootItem()

    def CollapseAll(self):
        self.tree.Collapse(self.tree.GetRootItem())

    def OnCollapse(self, event):
        self.CollapseAll()

    def GetSelection(self):
        return self.tree.GetSelection()

    def GetSignalAddress(self):
#        print self.tree.GetItemText(self.tree.GetItemParent(self.tree.GetSelection())) + self.tree.GetItemText(self.tree.GetSelection())
        return str(self.tree.GetItemText(self.tree.GetItemParent(self.tree.GetSelection())) + self.tree.GetItemText(self.tree.GetSelection()))

    def GetItemText(self, item):
        if item:
            return self.tree.GetItemText(item)
        else:
            return ""

    def OnItemExpanded(self, evt):
        pass

    def OnItemCollapsed(self, evt):
        pass

    # HACK: better to store a dict{item_name: item_id} ??
    def GetItemByLabel(self, search_text, root_item):
        # receives as search_text a full libmapper address (ex: /device/signal)
        item, cookie = self.tree.GetFirstChild(root_item)
        while item.IsOk():
            text = self.tree.GetItemText(item)
            if text.lower() == search_text.lower():
                return item
            if self.tree.ItemHasChildren(item) and len(search_text.split("/")) > 2:
                string = "/"+"/".join(search_text.split("/")[2:])
                match = self.GetItemByLabel(string, item)
                # if parent is the device, since we can have same signal name for different devices
                if match.IsOk() and self.tree.GetItemText(self.tree.GetItemParent(match)) == "/"+search_text.split("/")[1]:
                    return match
            item, cookie = self.tree.GetNextChild(root_item, cookie)
        return wx.TreeItemId()


Resources/storage.py
import json
import mapper

class Storage:

    def __init__(self):
        pass

    # This code is based on webmapper code (https://github.com/radarsat1/webmapper)
    def serialise(self, monitor, devices_list):
        sources = {}
        destinations = {}
        connections = {}
        new_connections = []
        next_src = 0
        next_dest = 0
        next_connection = 0
        modeStr = {mapper.MO_BYPASS: 'bypass',
                   mapper.MO_LINEAR: 'linear',
                   mapper.MO_CALIBRATE: 'calibrate',
                   mapper.MO_EXPRESSION: 'expression'}
        boundStr = {mapper.BA_NONE: 'none',
                   mapper.BA_MUTE: 'mute',
                   mapper.BA_CLAMP: 'clamp',
                   mapper.BA_FOLD: 'fold',
                   mapper.BA_WRAP: 'wrap'}

        for device in devices_list:
            for c in monitor.db.connections_by_device_name(device):
                this_connection = {
                  'src': [ c['src_name'] ],
                  'dest': [ c['dest_name'] ],
                  'mute': c['muted'],
                  'mode': modeStr[c['mode']],
                  'srcMin': c['src_min'],
                  'srcMax': c['src_max'],
                  'destMin': c['dest_min'],
                  'destMax': c['dest_max'],
                  'expression': c['expression'],
                  'boundMin': boundStr[c['bound_min']],
                  'boundMax': boundStr[c['bound_max']]
                }
                # To get proper expression nomenclature
                # dest[0] = src[0] NOT y = x
                this_connection['expression'] = this_connection['expression'].replace('y', 'dest[0]').replace('x', 'src[0]')
                new_connections.append(this_connection);

        contents = {"fileversion": "2.1", "mapping": {
                            "connections": new_connections
                            }
                    }
        return json.dumps(contents, indent=4)

    def deserialise(self, monitor, mapping_json, devices):
        js = json.loads(mapping_json)

        #The version we're currently working with
        version = '';
        if 'fileversion' in js:
            version = js['fileversion']
        elif 'mapping' in js and 'fileversion' in js['mapping']:
            version = js['mapping']['fileversion']

        modeIdx = {'bypass': mapper.MO_BYPASS,
                   'linear': mapper.MO_LINEAR,
                   'calibrate': mapper.MO_CALIBRATE,
                   'expression': mapper.MO_EXPRESSION}
        boundIdx = {'none': mapper.BA_NONE,
                    'mute': mapper.BA_MUTE,
                    'clamp': mapper.BA_CLAMP,
                    'fold': mapper.BA_FOLD,
                    'wrap': mapper.BA_WRAP}

        m = js['mapping']

        # sources and destinations are lists, devices are split from the second '/' character
        srcdevs = devices['sources']
        destdevs = devices['destinations']
        links = [( str(x), str(y) ) for x in srcdevs for y in destdevs]
        for l in links:
            monitor.link(l[0], l[1], {})
            monitor.poll(100)
        for c in m['connections']:
            #The name of the source signal (without device, assuming 1 to 1 for now)
            srcsig = str(c['src'][0]).split('/')[2]
            #And the destination
            destsig = str(c['dest'][0]).split('/')[2]

            for l in links:
               if monitor.db.get_link_by_src_dest_names(l[0], l[1]):
                    args = (str(l[0]+'/'+srcsig),
                            str(l[1]+'/'+destsig),
                            {})
                    if 'mode' in c:
                        args[2]['mode'] = modeIdx[c['mode']]
                    if 'expression' in c:
                        args[2]['expression'] = str(c['expression']
                                                    .replace('src[0]', 'x')
                                                    .replace('dest[0]', 'y'))
                    if 'srcMin' in c:
                        args[2]['src_min'] = c['srcMin']
                    if 'srcMax' in c:
                        args[2]['src_max'] = c['srcMax']
                    if 'destMin' in c:
                        args[2]['dest_min'] = c['destMin']
                    if 'destMax' in c:
                        args[2]['dest_max'] = c['destMax']
                    if 'boundMin' in c:
                        args[2]['bound_min'] = boundIdx[c['boundMin']]
                    if 'boundMax' in c:
                        args[2]['bound_max'] = boundIdx[c['boundMax']]
                    if 'mute' in c:
                        args[2]['muted'] = c['mute']

                    # If connection already exists, use 'modify', otherwise 'connect'.
                    # Assumes 1 to 1, again
                    cs = list(monitor.db.connections_by_device_and_signal_names(
                        (l[0]).split('/')[1], srcsig,
                        (l[1]).split('/')[1], destsig) )
                    if len(cs) > 0:
                        monitor.modify_connection(args[0], args[1], args[2])
                    else:
                        monitor.connect(*args)

if __name__ == "__main__":
    m = mapper.monitor(autosubscribe_flags=mapper.SUB_DEVICE)
    m.request_devices()
    m.poll(100)
    devices = [i["name"] for i in m.db.all_devices()]

    s = Storage()

    for d in devices:
        print s.serialise(m, d)



Resources/mymapper.py
import mapper

class MyMapper():
    def __init__(self):
        self.mon = mapper.monitor(autosubscribe_flags=mapper.SUB_DEVICE)
        self.initMonitor()
        self.mo_calibrate = mapper.MO_CALIBRATE
        self.mo_expression = mapper.MO_EXPRESSION
        self.mo_reverse = mapper.MO_REVERSE
        self.mo_linear = mapper.MO_LINEAR
        self.mo_bypass = mapper.MO_BYPASS

        self.ba_none = mapper.BA_NONE
        self.ba_mute = mapper.BA_MUTE
        self.ba_clamp = mapper.BA_CLAMP
        self.ba_fold = mapper.BA_FOLD
        self.ba_wrap = mapper.BA_WRAP

        self.modes_dict = {'Bypass': mapper.MO_BYPASS, 'Calibrate': mapper.MO_CALIBRATE, 'Reverse': mapper.MO_REVERSE,
                           'Linear': mapper.MO_LINEAR, 'Expression': mapper.MO_EXPRESSION}

        self.boundaries_dict = {'None': mapper.BA_NONE, 'Mute': mapper.BA_MUTE, 'Clamp': mapper.BA_CLAMP,
                           'Fold': mapper.BA_FOLD, 'Wrap': mapper.BA_WRAP}

        # we often need to fetch reversed info from the dictionaries above:
        self.modes_dict_rev = dict((v,k) for k, v in self.modes_dict.iteritems())
        self.boundaries_dict_rev = dict((v,k) for k, v in self.boundaries_dict.iteritems())

        self.OSC_devices = []

    def initMonitor(self):
        self.mon.request_devices()

    def poll(self, time):
        self.mon.poll(time)

    def setNetworkInterface(self, iface):
        self.admin = mapper.admin(iface=iface)
        self.mon = mapper.monitor(self.admin, autorequest=0)
        self.initMonitor()

    def setLink(self, source, dest, options={}):
        self.mon.link(source, dest, options)

    def Connect(self, src, dest, options=None):
        if not options: options = dict(mode=mapper.MO_LINEAR, bound_min=mapper.BA_WRAP,
                                       bound_max=mapper.BA_CLAMP)
        self.mon.poll(20)
        self.mon.connect(src, dest, options)

    def Disconnect(self, src, dest):
        self.mon.disconnect(src, dest)

    def Link(self):
        pass

    def Unlink(self):
        pass

    def getInputOutputDevices(self):
        "This is needed by storage code. Should return {sources: [/dev1, /dev2, /etc], destinations:[/dev3, /dev1]}"
        sources = []
        destinations = []
        devices = self.getAllDevices()
        for d in devices:
            if len(self.getInputsFromDevice(d)) != 0: # if device contains at least one input
                destinations.append(d)
            if len(self.getOutputsFromDevice(d)) != 0: # if device contains at least one output
                sources.append(d)
        in_out_devices = {'sources': sources, 'destinations': destinations}
        print in_out_devices
        return in_out_devices

    def Modify(self, src, dest, options=None):
        self.mon.modify_connection(src, dest, options)
        self.mon.poll(5)

    def getConnectionBySignalFullNames(self, src, dest):
        self.mon.poll(5)
        return self.mon.db.get_connection_by_signal_full_names(src, dest)

    def getSignalObjectBySignalFullName(self, dest, which):
        device = "/" + dest.split("/")[1]
        signal = "/" + "/".join(dest.split("/")[2:])
        if which == "input":
            return self.mon.db.get_input_by_device_and_signal_name(device, signal)
        elif which == "output":
            return self.mon.db.get_output_by_device_and_signal_name(device, signal)

    def getConnections(self):
        self.mon.poll(50)
        self.connections = [i for i in self.mon.db.all_connections()]
        return self.connections

    def getAllDevices(self):
        self.mon.poll(50)
        self.devices = [i["name"] for i in self.mon.db.all_devices()]
        return self.devices

    def getInputsFromDevice(self, device_name):
        self.getAllDevices()
        inputs_list = [i for i in self.mon.db.inputs_by_device_name(device_name)]
        return inputs_list

    def getOutputsFromDevice(self, device_name):
        self.getAllDevices()
        outputs_list = [i for i in self.mon.db.outputs_by_device_name(device_name)]
        return outputs_list

    def createDevice(self, device_name):
        self.OSC_dev = mapper.device(device_name)
        self.OSC_dev.add_output("/x", 1, 'f', None, -1000, 1000)
        self.OSC_dev.poll(100)
        self.OSC_devices.append(self.OSC_dev)
        return self.OSC_dev

# petit debug
if __name__ == "__main__":
    m = MyMapper()
    #   print m.getAllDevices()
    #   print m.getInputsFromDevice('/om.1')
    #   print m.getOutputsFromDevice('/om.1')
    #   print m.Connect("/om.1", "/om.2")
    print m.getConnections()
